marahgh/assignment2/BinaryTreeAncestors.h

```diff
@@ -0,0 +1,179 @@
+#include <iostream>
+#include <vector>
+#include <algorithm>
+
+using namespace std;
+/*The Binary Tree class,it has no duplicate values */
+template <typename S>
+class BinaryTree {
+  /*Node's Tree class*/
+  template <typename T>
+  class Node {
+   private:
+    T data;
+    Node<T>* left;
+    Node<T>* right;
+    friend class BinaryTree;
+
+   public:
+    /*Node's constructor */
+    Node(T data) : data(data), left(nullptr), right(nullptr) {}
+
+    /*Inserts a node as the right son of this node*/
+    bool insertRight(Node<T>* node) {
+      if (right) {
+        return false;
+      }
+      right = node;
+      return true;
+    }
+
+    /*Inserts a node as the right son of this node*/
+    bool insertLeft(Node<T>* node) {
+      if (left) {
+        return false;
+      }
+      left = node;
+      return true;
+    }
+  };
+
+  /****************************************************************************/
+  /*Binary Tree fields: the root node of the tree*/
+  Node<S>* root;
+
+  /****************************************************************************/
+
+  /*Deleting a binary Tree recursive auxiliary function*/
+  void recursiveDelete(Node<S>* root) {
+    if (!root) {
+      return;
+    }
+    recursiveDelete(root->left);
+    recursiveDelete(root->right);
+    delete root;
+  }
+
+  /*an auxiliary insert recursive function of the insert method.
+  returns true if insertion succeeded, flase otherwise.
+  leftOrRight is string which can be either right or left */
+  bool insertElement(Node<S>* treeRoot, Node<S>* node, S parent,

```

Comment:
Using string for leftOrRight is questionable. bool is an option, but in general bool parameters are easy to use wrong.
Another option is an enum.